File: lisp.rb
#!/usr/bin/env ruby
class Atom
  attr_reader :name
  def initialize(name)
    @name = name
  end
  def to_s()
    @name
  end
end

#monkey patch
class Cons
  include Enumerable
  attr_accessor :car
  attr_accessor :cdr
  def initialize(car, cdr)
    @car = car
    @cdr = cdr
  end
  def append(obj)
    li = self
    while li.cdr
      li = li.cdr
    end
    li.cdr = Cons.new(obj, nil)
  end
  def each()
    li = self
    while li
      yield li.car
      li = li.cdr
    end
  end    
  def to_s()
    (not @car and not @cdr) ? '()' : "(#{to_a.join(' ')})"
  end
end

class Func
  attr_reader :func
  def initialize(func)
    @func = func
  end
  def to_s()
    '%func'
  end
end

class Lambda
  attr_reader :args,:exp
  def initialize(args, exp)
    @args = args
    @exp = exp
  end
  def to_s()
    "#{@args} #{@exp}"
  end
end

module Enumerable
  def to_cons()
    raise 'to_cons for empty array' unless length > 0
    l = Cons.new(self[0], nil)
    shift
    each do |i|
      l.append i
    end
    l
  end
end


class Lisp
  def init_env()
    @tee = Atom.new '#T'
    @none = Cons.new nil,nil

    %w(QUOTE CAR CDR CONS EQUAL ATOM COND LAMBDA LABEL).map{|name|
      [Atom.new(name.upcase),Func.new(self.method('fn_'+name.downcase))].to_cons
    }.to_cons
  end

  def fn_car(args, env)
    args.car.car
  end
  def fn_cdr(args, env)
    args.car.cdr
  end
  def fn_quote(args, env)
    args.car
  end
  def fn_cons(args, env)
    l = Cons.new(args.car, nil)
    args = args.cdr.car
    args.each do |i|
      l.append i
    end
    l
  end
  def fn_equal(args, env)
    first = args.car
    second = args.cdr.car
    r = ((first.name==second.name) ? @tee : @none)
    r
  end
  def fn_atom(args, env)
    args.car.is_a?(Atom) ? @tee : @none
  end
  def fn_cond(args, env)
    args.each do |i|
      pred = eval i.car,env
      ret = i.cdr.car
      return eval(ret,env) if pred != @none
    end
    return @none
  end

  def fn_lambda(args, env)
    lambda = args.car
    args = args.cdr
    li = interleave(lambda.args, args)
    exp = replace_atom(lambda.exp, li)
    eval(exp,env)
  end

  def fn_label(args, env)
    env.append [Atom.new(args.car.name), args.cdr.car].to_cons
    @tee
  end

  def interleave(c1, c2)
    c1.to_a.zip(c2.to_a).map {|i| i.to_cons}.to_cons
  end

  def replace_atom(exp, to)
    if exp.is_a? Cons
      exp.map {|i| replace_atom i,to}.to_cons
    else
      to.each {|i|
        atom = i.car
        replacement = i.cdr.car
        return replacement if atom.name == exp.name
      }
      exp
    end
  end

  def lookup(name, env)
    env.each do |i|
      nm = i.car
      val = i.cdr.car
      return val if nm.name == name
    end
    nil
  end

  def eval_fn(exp, env)
    symbol = exp.car
    args = exp.cdr
    if symbol.is_a? Lambda
      fn_lambda(exp,env)
    elsif symbol.is_a? Func
      symbol.func.call(args,env)
    else
      exp
    end
  end

  def eval(exp, env)
    if exp.is_a? Cons
      if exp.car.is_a?(Atom) and exp.car.name=='LAMBDA'
        largs = exp.cdr.car
        lexp = exp.cdr.cdr.car
        Lambda.new largs,lexp
      else
        eval_fn(exp.map {|e| eval(e,env)}.to_cons, env)
      end
    else
      lookup(exp.name, env) || exp
    end
  end

  def next_token(input)
    ch = input.getc
    while ch and ch.chr =~ /\s/
      ch = input.getc
    end
    
    exit(0) if not ch
    ch = input.getc if ch.chr=='\n'
    return Atom.new(')') if ch.chr==')'
    return Atom.new('(') if ch.chr=='('
    
    name = ""
    while not ch.chr=~/\s/ and ch.chr!=')'
      name += ch.chr
      ch = input.getc
    end
    
    input.ungetc ch if ch.chr==')'
    
    Atom.new(name)
  end

  def read_tail(input)
    token = next_token(input)
    
    if token.name==')'
      nil
    elsif token.name=='('
      first = read_tail(input)
      second = read_tail(input)
      Cons.new(first, second)
    else
      first = token
      second = read_tail(input)
      Cons.new(first, second)
    end
  end

  def read(f)
    token = next_token(f)
    token.name=='(' ? read_tail(f) : token
  end

  def main()
    env = init_env()
    if ARGV.length > 0
      $stdout.flush
      f = File.open ARGV[0]
    else
      f = $stdin
    end
    
    while 1
      print '> '
      $stdout.flush
      r = read(f)
      e = eval(r, env)
      puts e
    end
  end

end

Lisp.new.main

File: lisp.py
#!/usr/bin/env python

import sys

class Atom(object):
    def __init__(self, name):
        self.name = name
    def __str__(self):
        return self.name
#    def __eq__(self,other):
#        return isinstance(other,Atom) and self.name==other.name
class Cons(object):
    def __init__(self, car, cdr):
        self.car = car
        self.cdr = cdr # cdr must be None or Cons
    def append(self, obj):
        li = self
        while li.cdr:
            li = li.cdr
        assert(not li.cdr)
        li.cdr = Cons(obj, None)
    def to_list(self):
        l = []
        li = self
        while li:
            l.append(li.car)
            li = li.cdr
        return l
    def __str__(self):
        if not self.car and not self.cdr:
            return '()'
        return '(%s)'%' '.join([str(e) for e in self.to_list()])
class Func(object):
    def __init__(self, func):
        self.func = func
    def __str__(self):
        return '%func'
class Lambda(object):
    def __init__(self, args, exp):
        self.args = args
        self.exp = exp
    def __str__(self):
        return '#%s %s'%(self.args, self.exp)

# primitives
tee = Atom('#T')
nil = Cons(None,None)

def fn_car(args, env):
    return args.car.car
def fn_cdr(args, env):
    return args.car.cdr
def fn_quote(args, env):
    return args.car
def fn_cons(args, env):
    l = Cons(args.car, None)
    args = args.cdr.car
    for i in args.to_list():
        l.append(i) 
    return l
def fn_equal(args, env):
    first = args.car
    second = args.cdr.car
    if first.name==second.name:
        return tee
    else:
        return nil
def fn_atom(args, env):
    if isinstance(args.car,Atom):
        return tee
    else:
        return nil
def fn_cond(args, env):
    for i in args.to_list():
        pred = eval(i.car, env)
        ret = i.cdr.car
        
        if pred != nil:
            return eval(ret,env)
    return nil

def interleave(c1, c2):
    l = Cons(Cons(c1.car, Cons(c2.car, None)), None)
    c1 = c1.cdr
    c2 = c2.cdr
    
    for i in c1.to_list():
        l.append( Cons(c1.car, Cons(c2.car,None)) )
        c2 = c2.cdr
    return l

def to_cons(li):
    l = Cons(li[0],None)
    for i in li[1:]:
        l.append(i)
    return l
def replace_atom(exp, to):
    if isinstance(exp, Cons):
        return to_cons([replace_atom(i,to) for i in exp.to_list()])
    else:
        assert isinstance(exp, Atom)
        for item in to.to_list():
            atom = item.car
            replacement = item.cdr.car
            
            if atom.name==exp.name:
                return replacement
        return exp

def fn_lambda(args, env):
    lamb = args.car
    args = args.cdr
    assert isinstance(lamb,Lambda)
    li = interleave(lamb.args,args)
    exp = replace_atom(lamb.exp,li)
    return eval(exp,env)

def fn_label(args, env):
    env.append(Cons(Atom(args.car.name), Cons(args.cdr.car, None)))
    return tee

def lookup(name, env):
    for item in env.to_list():
        nm = item.car
        val = item.cdr.car
        if nm.name == name:
            return val

def init_env():
    def entry(name, fn):
        return Cons(Atom(name),Cons(Func(fn),None))
    env = Cons(entry('QUOTE',fn_quote),None)
    env.append(entry('CAR',fn_car))
    env.append(entry('CDR',fn_cdr))
    env.append(entry('CONS',fn_cons))
    env.append(entry('EQUAL',fn_equal))
    env.append(entry('ATOM',fn_atom))
    env.append(entry('COND',fn_cond))
    env.append(entry('LAMBDA',fn_lambda))
    env.append(entry('LABEL',fn_label))

    return env

def eval_fn(exp, env):
    symbol = exp.car
    args = exp.cdr
    
    if isinstance(symbol, Lambda):
        return fn_lambda(exp,env)
    elif isinstance(symbol, Func):
        return symbol.func(args,env)
    else:
        return exp

def eval(exp, env):
    #print 'eval: %s'%exp
    if isinstance(exp, Cons):
        if isinstance(exp.car, Atom) and exp.car.name=='LAMBDA':
            largs = exp.cdr.car
            lexp = exp.cdr.cdr.car
            return Lambda(largs,lexp)
        else:
            return eval_fn(to_cons([eval(e,env) for e in exp.to_list()]), env)
    else:
        assert isinstance(exp, Atom)
        val = lookup(exp.name, env)
        if val:
            return val
        return exp

def next_token(input):
    ch = input.getc()
    while ch and ch.isspace():
        ch = input.getc()
        
    if ch=='\n':
        ch = input.getc()
    if not ch:
        exit(0)
    
    if ch==')':
        return Atom(')')
    if ch=='(':
        return Atom('(')
    
    buffer = ''
    index = 0
    while not ch.isspace() and ch!=')':
        buffer += str(ch)
        ch = input.getc()
        
    if ch==')':
        input.ungetc(ch)

    return Atom(buffer)

def read_tail(input):
    token = next_token(input)
    
    if token.name==')':
        return None
    elif token.name=='(':
        first = read_tail(input)
        second = read_tail(input)
        return Cons(first,second)
    else:
        first = token
        second = read_tail(input)
        return Cons(first,second)

def read(f):
    token = next_token(f)
    if token.name=='(':
        return read_tail(f)
    return token

class FileBuffer:
    def __init__(self,fileobj):
        self.f = fileobj
        self.buffer = []
    def eof(self):
        if self.buffer:
            return False
        c = self.getc()
        if not c:
            return True
        self.ungetc(c)
        return False

    def getc(self):
        if self.buffer:
            self.buffer,r = self.buffer[:-1], self.buffer[-1]
            return r
        c = self.f.read(1)
        if c=='': return None
        return c

    def ungetc(self,char):
        self.buffer.append(char)

if __name__=='__main__':
    env = init_env()
    
    if len(sys.argv)>1:
        f = FileBuffer(open(sys.argv[1],'r'))
    else:
        f = FileBuffer(sys.stdin)

    while not f.eof():
        sys.stdout.write('> ')
        sys.stdout.flush()
        r = read(f)
        e = eval(r, env)
        print e

